File: 3_longest_substring_without_repeating_characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        left = 0
        right = 0
        length = 0
        cache = {}
        
        while right < len(s):
            if s[right] in cache and cache[s[right]] >= left:
                left = cache[s[right]] + 1
            else:
                cache[s[right]] = right
                right += 1
                
            if right-left > length:
                length = right-left
            
        return length

File: c++/3.longest_substring_without_repeating_characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        int left = 0;
        int right = 0;
        int maxlen = 0;
        unordered_map<int, int> cache;
        
        while(right < s.size()){
            if(cache.find(s[right]) != cache.end() && cache[s[right]] >= left){
                left = cache[s[right]] + 1;
                
            }else{
                cache[s[right]] = right;
                right += 1;
            }
            maxlen = max(right-left, maxlen);
        }
        return maxlen;
        
    }
};